syntax = "proto3";

package obj_store;

import "auth.proto";

enum HashType {
    SHA256 = 0;
    SHA512 = 1;
    SHA3_256 = 2;
    SHA3_512 = 3;
};

service ObjStore{
  rpc Write(WriteRequest) returns (Result);
  rpc Read(ReadRequest) returns (ContentResult);
  rpc Hash(HashRequest) returns (ContentResult);
  rpc Delete(DeleteRequest) returns (Result);
};

message WriteRequest {
  auth.TaskHeader hdr = 1;
  bytes id = 2;
  repeated string paths = 3; // required when creating file
  int64 offset = 4;  // -1 = end of file
  bytes data = 5;
};

message ReadRequest {
  auth.TaskHeader hdr = 1;
  bytes id = 2;
  int64 offset = 3; // -1 = end of file
  int64 data_len = 4;
};

message HashRequest {
  auth.TaskHeader hdr = 1;
  bytes id = 2;
  HashType hash_type = 3;
  int64 offset = 4; // -1 = end of file
  int64 data_len = 5; // default = to file end
};

message DeleteRequest {
  auth.TaskHeader hdr = 1;
  bytes id = 2;
};

message Result { auth.OperationResult result = 1; };

message ContentResult {
  auth.OperationResult result = 1;
  bytes content = 2;
};
